modules/phase_field/src/kernels/AllenCahnElasticEnergyOffDiag.C

```diff
@@ -0,0 +1,56 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AllenCahnElasticEnergyOffDiag.h"
+#include "MathUtils.h"
+#include "RankTwoTensor.h"
+
+registerMooseObject("PhaseFieldApp", AllenCahnElasticEnergyOffDiag);
+
+template <>
+InputParameters
+validParams<AllenCahnElasticEnergyOffDiag>()
+{
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("This kernel calculates off-diagonal jacobian of elastic energy in "
+                             "AllenCahn respect to displacements");
+  params.addCoupledVar("displacements",
+                       "The string of displacements suitable for the problem statement");
+  params.addParam<MaterialPropertyName>(
+      "F_name", "E_el", "Name of material property storing the elastic energy");
+  params.addParam<MaterialPropertyName>("mob_name", "L", "The mobility used with the kernel");
```

Comment:
I'd like to move away from naming material property parameters `*_name`. The user can supply a plain number here, not just a name.